pupil-spa/src/app/practice-question/practice-question.component.ts

```diff
@@ -0,0 +1,238 @@
+  styleUrls: ['../question/question.component.css']
+})
+export class PracticeQuestionComponent implements OnInit, AfterViewInit {
+
+  /**
+   * The time in ms since the epoch when the answer will be automatically submitted.
+   * Used to calculate the remaining time counter.
+   */
+  protected stopTime: number;
+
+  /**
+   * Flag to indicate that the answer has been submitted (either manually or on timeout)
+   */
+  protected submitted = false;
+
+  /**
+   * Store the return value of setTimeout used to submit the answer.  The setTimeout() is cancelled
+   * if the user manually submits the answer.
+   */
+  protected timeout: number;
+
+  /**
+   * Store the return value of setInterval - so it can be cancelled
+   */
+  protected countdownInterval: number;
+
+  /**
+   * Experimental: Flag to indicate whether we want speech synthesis
+   */
+  protected hasSpeechSynthesis = false;
+
+  /**
+   * Reference to global window object
+   */
+  protected window: any;
+
+  /**
+   * The users answer made up of recorded numbers.
+   * This is a string as the user may have leading zeros which we want to store.
+   * Used in the template.
+   */
+  public answer = '';
+
+  /**
+   * The remaining time in seconds until the answer is automatically submitted
+   * Used in the template.
+   */
+  public remainingTime: number;
+
+  @Input() public factor1 = 0;
+
+  @Input() public factor2 = 0;
+
+  @Input() public questionTimeoutSecs;
+
+  @Output() public manualSubmitEvent: EventEmitter<any> = new EventEmitter();
+
+  @Output() public timeoutEvent: EventEmitter<any> = new EventEmitter();
+
+  constructor(protected auditService: AuditService,
+              protected windowRefService: WindowRefService) {
+    this.window = windowRefService.nativeWindow;
+  }
+
+  ngOnInit() {
+    this.remainingTime = this.questionTimeoutSecs;
+  }
+
+  /**
+   * Start the timer when the view is ready.
+   */
+  ngAfterViewInit() {
+    this.auditService.addEntry(new QuestionRendered());
+    // Start the countdown and page timeout timers
+    this.startTimer();
+  }
+
+  /**
+   * Start the countdown timer on the page and set the time-out counter
+   */
+  startTimer() {
+    this.stopTime = (new Date().getTime() + (this.questionTimeoutSecs * 1000));
+
+    // Set the amount of time the user can have on the question
+    this.timeout = this.window.setTimeout(() => {
+      this.sendTimeoutEvent();
+    }, this.questionTimeoutSecs * 1000);
+
+    // Set the countdown timer on the page
+    this.countdownInterval = this.window.setInterval(() => {
+      let timeLeft = (this.stopTime - (new Date().getTime())) / 1000;
+      if (timeLeft < 0) {
+        clearInterval(this.countdownInterval);
+        timeLeft = 0;
+      }
+      this.remainingTime = Math.ceil(timeLeft);
+    }, 100);
+  }
+
+  /**
+   * Check a manual submission to see if it is allowed.
+   * @return {boolean}
+   */
+  answerIsLongEnoughToManuallySubmit() {
```

Comment:
A bit confusing naming convention in my opinion.
Shouldn't that be **hasAnswer**?